XenAdmin/Diagnostics/Checks/ServerSelectionCheck.cs

```diff
@@ -51,16 +54,47 @@ public ServerSelectionCheck(Pool pool, Pool_update update, List<Host> selectedSe
             this.pool = pool;
             this.update = update;
             this.selectedServers = selectedServers;
+            isNewVersionUpdate = false;
+        }
+
+        public ServerSelectionCheck(Pool pool, XenServerPatchAlert alert, List<Host> selectedServers)
+            : base(Helpers.GetMaster(pool.Connection))
+        {
+            this.pool = pool;
+            this.alert = alert;
+            this.selectedServers = selectedServers;
+            isNewVersionUpdate = true;
         }
 
         protected override Problem RunCheck()
         {
             if (!Host.IsLive())
                 return new HostNotLiveWarning(this, Host);
 
+            if (isNewVersionUpdate)
+            {
+                //Must select all hosts in a homogeneous pool
+                if (pool.IsPoolFullyUpgraded() && pool.Connection.Cache.Hosts.Any(h => !selectedServers.Contains(h)))

```

Comment:
Suggestion (to avoid multiple IsPoolFullyUpgraded calls):
```
if (pool.IsPoolFullyUpgraded())
{
     //Must select all hosts in a homogeneous pool
    if (pool.Connection.Cache.Hosts.Any(h => !selectedServers.Contains(h)))
        return new ServerSelectionProblem(this, pool);
}
else
{
    //For mixed pool, precheck blocks update to a new version which is higher than master without selecting it
    //Otherwise, skip the precheck for a mixed pool and issue warning. Because the version update may not be compatible to all hosts.
    if (Helpers.productVersionCompare(alert.NewServerVersion.Version.ToString(), Host.ProductVersion()) > 0 && !selectedServers.Contains(Host))
         return new MasterVersionNotCompatibleProblem(this, pool);
    else
        return new MixedPoolServerSelectionWarning(this, pool);
}
```